### ISSUE TITLE
MAINT/DOC: Pre-render notebook HTML for faster loading

### ISSUE THREAD

**Issue by adriangb** [adriangb/scikeras#130]
Crossref https://github.com/adriangb/scikeras/pull/123#discussion_r524840156

**Comment by adriangb** [comment]
@stsievert it seems like GitHub automatically caches notebook rendering.

I was looking at [this](https://github.com/adriangb/scikeras/blob/whole-dataset-transformer/notebooks/DataTransformers.ipynb) notebook, and the first time it loaded it took ~10 seconds. After that, it loaded instantly, even in another browser, logged out of my account. I'm not sure how long they cache it for though.

Can you check if it loads quickly for you?

**Comment by stsievert** [comment]
In the Opera browser (which relies on the Chromium):

* First time: waited 13s, got message "something went wrong. Try again?"
* Second time: 2–3s.
* Subsequent times in different browsers: between 2–3s.

There's a variety of tools to support static HTML rendering notebooks in Sphinx docs, including [nbsphinx], [sphinxcontrib-jupyter] and (possibly) [jupyter-sphinx].

[nbsphinx]:https://nbsphinx.readthedocs.io/
[sphinxcontrib-jupyter]:https://sphinxcontrib-jupyter.readthedocs.io/
[jupyter-sphinx]:https://jupyter-sphinx.readthedocs.io/en/latest/

**Comment by adriangb** [comment]
Interesting. I'm not sure how they do their caching...

Aside from pre-rendering, another option may be to just link only Google Colab and not GitHub.

Another option I've considered is to move to documentation like [FastAPIs](https://fastapi.tiangolo.com/), where the code is in separate Python files that are embedded into the docs, and that also get tested as part of the library's tests.

**Comment by stsievert** [comment]
> and that also get tested as part of the library's tests.

nbsphinx runs the notebooks every time you build the documentation (as long as the output cells are empty). I'm sure you can integrate this into your CI runs, and empty output cells would make the diffs a lot easier to work with on GitHub.

**Comment by adriangb** [comment]
I would like to reduce the diffs for notebooks. I will look into it, thank you for the continued recommendations.

**Comment by adriangb** [comment]
I implemented a docs overhaul in #174 . I initially tried to implement something simple like https://github.com/adriangb/scikeras/issues/130#issuecomment-765541036, but I ran into issues because RTDs workers kept crashing due to the resource usage of our notebooks. So I arrived at another solution that basically consists of:
- Storing the notebooks as markdown so that the diffs are kept clean.
- In GitHub actions, run all notebooks and build sphinx docs.
- Deploy the fully rendered docs to the `docs-deploy` branch, which is served via GitHub pages.
This includes building preview docs for PRs.
